src/Model/ModelData.php

```diff
@@ -305,12 +306,18 @@ public function exists(): bool
         return true;
     }
 
+    public function __toString(): string
+    {
+        return $this->forTemplate();
```

Comment:
Won't this risk kind of drastically changing somethings rather negatively, possibly some parts of the debugging experience?

We exactly are we calling forTemplate() in __toString()? This just seems wrong somehow, like it's a misuse of __toString()?

Reply:
> Won't this risk kind of drastically changing somethings rather negatively, possibly some parts of the debugging experience?

If someone is relying on `(string) $model` instead of `get_class($model)` they're relying on very flaky behaviour which we already don't respect in `DBField` which does exactly what this method is doing.
Doing this here accomplishes the following:

1. Stops people relying on string casting to check the class name, which is flaky at best and breaks down as soon as a subclass wants to return something else from this method
2. Makes this method's usage more consistent across subclasses - it's taking what was already happening in `DBField` and making it standard for _all_ models.

> We exactly are we calling forTemplate() in __toString()? This just seems wrong somehow, like it's a misuse of __toString()?

https://www.php.net/manual/en/stringable.tostring.php "Returns the string representation of the object."

This method is specifically for this purpose. The rendered markup of a model _is_ the string representation of that model.

This is what other template rendering engines will call to try to get the textual markup that should be injected into templates.

What do you think this method should be used for instead of this?

Reply:
> If someone is relying on (string) $model instead of get_class($model) they're relying on very flaky behaviour which we already don't respect in DBField which does exactly what this method is doing.

I'm not sure how that's relevant. By itself that's not a good reason to implement a magic method here.

> it's taking what was already happening in DBField and making it standard for all models.

Having a look at that code, while that does call DBField::forTemplate(), ultimately all that's doing is calling DBField::getValue() which won't be rendering any HTML

> The rendered markup of a model is the string representation of that model.

I'm not sure that is correct? While that makes sense in a template context, ModelData is not a class that's exclusively used in templates. Wasn't that why we renamed ViewableData to ModelData, as lots of ViewableData subclasses weren't exclusively used in a template context?

I might be making an incorrect assumption, though won't ModelData::forTemplate() be rendering HTML is some cases, which will be out of context and unhelpful some of the time?

I also just generally disklike magic methods as it's really not clear what's going on. Do we even need __toString() here? Can we not simply just update whatever "calls" thing to simply use ->forTemplate() instead?

Reply:
> I'm not sure how that's relevant. By itself that's not a good reason to implement a magic method here.

The method was _already implemented_. I'm just changing what it returns.

> Having a look at that code, while that does call DBField::forTemplate(), ultimately all that's doing is calling DBField::getValue() which won't be rendering any HTML

I don't understand your point here... is your problem with the implementation of `ModelData::__toString()` that it returns HTML markup?

> won't ModelData::forTemplate() be rendering HTML is some cases

Yes.

> which will be out of context and unhelpful some of the time?

In cases where that's not helpful, people probably shouldn't be casting the object to a string. You cast something as a string when you want the string representation of it, usually to present to the end user. In what case would you do that for `ModelData` outside of presenting it in the browser?

In an API, for example, you'd be sending individual fields, probably as JSON, not sending a string representation of the _whole_ model.

> I also just generally disklike magic methods

The method was already there, so if your problem is with the method _in general_  this isn't the place to discuss it.

Reply:
This doesn't affect the CMS 5 PRs in any case. So feel free to continue this discussion, but it shouldn't block those PRs from being merged.

Reply:
The abstraction I've done doesn't, in general, rely on this change to `__toString()`. I just wanted to make this method more standard across `ModelData` and dissuade people on relying on it for checking the class since that's a really bad way to check the class.

Reply:
Can we revert it to how it was then? Whether or not it's a good idea, it seems like it's just entirely unrelated to this PR and it's kind of just derailing the actual peer review. Could also please remove all of the added `implements Stringable` stuff too, they don't belong in the PR either. My preference for those would be to get rid of the __toString() methods though that's a discussion for an entirely different card

Reply:
It is very clearly [best practice](https://www.php.net/manual/en/language.oop5.magic.php#object.tostring) to explicitly implement the `Stringable` interface when using `__toString()`.
It is also best practice for those methods to be explicitly typed to return a string.
It is also very clearly _bad_ practice to rely on casting an object to string to get the class of that object.

My changes were addressing all of these.

Despite that, I'll revert those changes so that we can move on.